Adds new `no-unused-prop-types` rule

EvNaverniouk opened the pull request:
Fixes #226 by adding an new `unused-prop-types` rule.

The rule is based very heavily on the `prop-types` rule and shares a lot of the same code. For the time being, I have not made any attempts at breaking out the shared code into a reusable utility file. Please advise on the best way to do this. I was simply thinking of dropping a `proptypes.js` file into the `utils` directory with all the shared functions. If that's acceptable, let me know and I will do that.

Have tested against several React projects with great success so far. No false positive problems yet, but it needs a lot more testing against other repos. I'm hoping other users will report issues as they try it. For my needs, this rule is working great.

Can be tested by following these steps:

```
npm install EvNaverniouk/eslint-plugin-react#6d309e13a65a1032577ae0c98081d778f85d515a
```

Then enable the rule in your eslintrc:

```
"rules": {
    "react/unused-prop-types": 2
}
```

EvNaverniouk commented:
I found one condition where this reports a false positive:

```
class Hello extends Component {
	static propTypes = {
		params: PropTypes.shape({
			id: PropTypes.string
		})
	}
	render () {
		const {params} = this.props;
		const id = (params || {}).id;
		return <span>{id}</span>;
	}
}
```

Not sure yet how to get `(params || {}).id` to get detected properly. Any advise would be welcome. I'll keep looking into it.

mattdell commented:
I'm going to have a look at this now but want to flag to @coveralls that the PR has conflicts. Can you fix @coveralls ?

I think the issue using the pattern `const {params} = this.props` is something we can live with. It would be better to find unused instances of `this.props.params` as a first pass and we can work on extending this for the former pattern as well.

mattdell commented:
I'm going to have a look at this now but want to flag to @EvNaverniouk that the PR has conflicts. Can you fix @EvNaverniouk ?

I think the issue using the pattern `const {params} = this.props` is something we can live with. It would be better to find unused instances of `this.props.params` as a first pass and we can work on extending this for the former pattern as well.

EvNaverniouk commented:
Merged master into this PR to resolve conflicts. Thanks @mattdell

ljharb commented:
@EvNaverniouk it'd be nice if you could *rebase* master instead, so there's no merge commits (and to squash all the "wip" commits)

mattdell commented:
There's an issue when using Atom and linter-eslint package where the linter failure is shown on the first line of code.

Here is an example:
![screen shot 2016-06-07 at 17 16 18](https://cloud.githubusercontent.com/assets/2536442/15865793/02db20f0-2cd4-11e6-95a7-b772745b95e5.png)
The console shows the correct output. Not sure if this is an issue with this repo or the linter-eslint package?

ljharb commented:
Seems like an issue with your atom config. If it works on the command line, it works, and all else is a local config issue.

mattdell commented:
You are correct @ljharb! A restart of Atom fixed that issue.

This looks great to me. Found 35 unused prop types!

EvNaverniouk commented:
@ljharb Rebased.

mattdell commented:
What needs to happen to get this merged?

antialias commented:
looks like it needs another merge / rebase

ljharb commented:
@EvNaverniouk would you mind rebasing this?

EvNaverniouk commented:
@ljharb Rebase completed.

ljharb commented:
ping @yannickcr :-)

Faradey27 commented:
@EvNaverniouk do you plan to fix conflicts?

antialias commented:
I resolved the conflicts between this branch and master and pushed them to my fork: https://github.com/antialias/eslint-plugin-react/commit/dffa1f341037f6b82bb02f9fee6542200f942eaa

There were no ambiguities, i.e. I'm 100% confident that I resolved them correctly.

@yannickcr, I could submit a replacement PR or @EvNaverniouk, you could hard reset this branch to the commit from my branch and the PR would show as resolved.

EvNaverniouk commented:
Resolved conflicts and rebased again.

EvNaverniouk commented:
Looks like some of my tests are failing after merging from latest master. Investigating...

EvNaverniouk commented:
Ok. Fixed the conflicts and rebased again. Also added a new commit to fix a bug in the `Component.js` utility method. I left that as a separate commit so you can clearly see it: https://github.com/yannickcr/eslint-plugin-react/pull/611/commits/fcd9d59271fbdf2874310eb2a226adf092fa6e4a

petersendidit commented on lib/rules/unused-prop-types.js:
```diff
@@ -0,0 +1,864 @@
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = Components.detect(function(context, components, utils) {
```
This should use the [new eslint rule format](http://eslint.org/blog/2016/07/eslint-new-rule-format)

EvNaverniouk commented:
@petersendidit Updated PR to use new ESLint rule format.

dhruska commented:
I pulled down this branch into one of my projects to give it a try, since I'm really excited about this - it worked great, except for in one component where it did not detect that I was only using a prop in `componentWillReceiveProps` on the `nextProps` object. Granted, it's probably a poor design pattern in my project that I am doing this, but the linter threw an error when I was indeed using the prop.

EvNaverniouk commented:
@dhruska Can you send me a copy of the code that reproduces that?

dhruska commented:
@EvNaverniouk Let me work on getting you an example. I tried to create a simple component that would reproduce the issue and was not able to.

burabure commented:
this will be very useful! I'll try to test this on our codebase (1M+ lines) and report back

lencioni commented on docs/rules/unused-prop-types.md:
```diff
@@ -0,0 +1,70 @@
+  }
+});
+
+function Hello({ name }) {
```
I don't understand why this is included in the examples--this component has no propTypes. Did you mean to include something like this?

```js
Hello.propTypes = {
  name: React.PropTypes.string.isRequired,
};
```

lencioni commented on docs/rules/unused-prop-types.md:
```diff
@@ -0,0 +1,70 @@
+
+* `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
+* `customValidators`: optional array of validators used for propTypes validation.
+* `skipShapeProps`: In some cases it is impossible to accurately detect whether or not a shape PropType's values are being used. Setting this option to true will skip validation of shape PropTypes.
```
I find the first sentence on this line difficult to grok. I think it would be clearer if it referenced `React.PropTypes.shape` specifically. e.g.

> In some cases it is impossible to accurately detect whether a `React.PropTypes.shape`'s values are being used.

Also, please include backticks around any words that are code references (e.g. `true`).

lencioni commented on docs/rules/unused-prop-types.md:
````diff
@@ -0,0 +1,70 @@
+```
+
+* `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
+* `customValidators`: optional array of validators used for propTypes validation.
````
It's not very clear to me what this option does exactly, when I might need it, or how to use it. Will you add a full example to the documentation?

EvNaverniouk replied:
To be honest with you, I don't know of good use cases for it either. I've included it only because the `prop-types` rule also has this option. And its documentation is exactly as this.

EvNaverniouk commented:
@lencioni Updated PR to clean up the docs as per your recommendations, fixed merge conflicts and rebased.

lencioni commented:
I think this looks good, thanks for all of your work on this rule. If this code is based off of the prop-types rule code, we currently have a bug in that rule where nested destructured props are not marked as used. (#296) It would be good to include some examples in your tests that demonstrate this.

EvNaverniouk commented:
@lencioni Not sure if that's what you had in mind, but I added a commented-out test at https://github.com/yannickcr/eslint-plugin-react/pull/611/commits/f083a13d13cf04331d69e76230d319248156aa13#diff-655c668b4b1964c5be4ff486d879ef45R1980 which can be enabled once #296 is fixed.

lencioni commented:
@EvNaverniouk Thanks for doing that! That's not quite what I had in mind. My point was more that if we are introducing code that has bugs that we are aware of, it would be nice to fix those bugs before merging. And actually, I think the bug might be the inverse in this rule (e.g. nested destructuring of a prop in propTypes is incorrectly flagged).

```jsx
 function Foo(props) {
   const { foo: { bar } } = props;
   return <div test={bar} />;
 }

 Foo.propTypes = {
   foo: PropTypes.shape({
     bar: PropTypes.number,
   }),
 };
```

Any chance you can add a test case for this, and if it fails, make it pass?

yannickcr commented:
I think this rule already adds a lot of value even if this bug is not fixed right now. So I am not against merging this rule with the bug and fix it later on (actually I still need to figure out how to fix it in the `prop-types` rule).

Also I'd prefer to name this rule `no-unused-prop-types` to match [ESLint Rule Naming Conventions](http://eslint.org/docs/developer-guide/working-with-rules#rule-naming-conventions).

yannickcr commented:
Ho, and thank you for this, this really great work ❤️

EvNaverniouk commented:
@lencioni To be honest with you, fixing that bug is a bit beyond my level of understanding of how the component and proptype detection works. However, I'm happy to merge the fix in into this rule once someone solves that bug for the `prop-types` rule. I've subscribed to it and will keep a close eye on it once a fix is found.

@yannickcr I've renamed the rule to `no-unused-prop-types` and rebased again.

lencioni commented:
@EvNaverniouk Perfect is the enemy of the good, so that sounds good to me. Thanks for your work on this!

yannickcr commented:
Merged, thanks!

Faradey27 commented:
thank you guys, nice rule and nice job

tristanbbq commented:
@yannickcr: I updated to 6.2.2 and added rule as following:
`"react/no-unused-prop-types": ["warn", { "customValidators": [], "skipShapeProps": true }],`
But I keep getting warnings `PropType is defined but props is never used`.

Something I'm missing here?

EvNaverniouk commented:
@tristanbbq Can you please open a new issue with a code sample I can use to reproduce?

tristanbbq commented:
@EvNaverniouk created https://github.com/yannickcr/eslint-plugin-react/issues/837
